src/proto/hybrid_vec.rs

```diff
@@ -0,0 +1,1832 @@
+
+    fn next(&mut self) -> Option<T> {
+        match *self {
+            IntoIter::Stack { ref array, ref mut curr, len } if *curr != len => { 
+                let ret = unsafe { ptr::read(&array[*curr]) };
+                *curr += 1;
+                Some(ret)
+            },
+            IntoIter::Heap { ptr: _, cap: _, ref mut curr, end } if *curr != end => unsafe {
+                if mem::size_of::<T>() == 0 {
+                    *curr = mem::transmute(*curr as usize + 1);
+                    return Some(ptr::read(mem::transmute(1us)));
+                }
+
+                let ret = ptr::read(*curr);
+                *curr = curr.offset(1);
+                Some(ret)
+            },
+            _ => None,
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let exact = match *self {
+            IntoIter::Stack { array: _, curr, len } => len - curr,
+            IntoIter::Heap { ptr: _, cap: _, curr, end } => {
+                let diff = (end as usize) - (curr as usize);
+                let size = mem::size_of::<T>();
+                diff / (if size == 0 {1} else {size})
+            }
+        };
+        
+        (exact, Some(exact))  
+    }
+}
+
+impl<T> DoubleEndedIterator for IntoIter<T> {
+    //#[inline]
+    fn next_back<'a>(&'a mut self) -> Option<T> {
+        match *self {
+            IntoIter::Stack { ref mut array, curr, ref mut len } if curr != *len => {
+                *len -= 1;
+                Some(unsafe { ptr::read(&array[*len]) })
+            },
+            IntoIter::Heap { ptr: _, cap: _, curr, ref mut end } if curr != *end => unsafe {
+                if mem::size_of::<T>() == 0 {
+                    *end = mem::transmute(*end as usize - 1);
+                    Some(ptr::read(mem::transmute(1us)))
+                } else {
+                    *end = end.offset(-1);
+                    Some(ptr::read(mem::transmute(*end)))
+                }
+            },
+            _ => None,
+        }
+    }
+}
+
+impl<T> ExactSizeIterator for IntoIter<T> {}
+
+impl<T> IntoIter<T> { 
+    pub fn into_inner(mut self) -> HybridVec<T> {
+        for _ in self {}
+
+        let ret = match self {
+            IntoIter::Stack { ref array, curr: _, len: _ } =>
+            HybridVec::Stack {
+                array: unsafe { ptr::read(array) },
+                len: 0,
+            },
+            IntoIter::Heap { ptr, cap, curr: _, end: _, } => unsafe { 
+                HybridVec::from_raw_parts(ptr, cap, 0) 
+            },
+        };
+
+        unsafe { mem::forget(self); }
+        ret
+    }    
+}
+
+#[unsafe_destructor]
+impl<T> Drop for IntoIter<T> {
+    fn drop(&mut self) {
+        for _ in *self {}
+
+        unsafe { 
+            let temp = mem::replace(self, mem::zeroed());
+            match temp {
+                IntoIter::Heap { ptr, cap, curr: _, end: _ } if cap != 0 => dealloc(ptr, cap),
+                _ => (),
+            }
+
+            mem::forget(temp);
```

Comment:
@huonw I considered adding a variant for registering completed drop but I figured I would keep running into the same problem since replacing it with `Heap { ptr: EMPTY, len: 0, cap: 0 }` didn't stop the drop glue from running.